src/CppParser/Parser.cpp

```diff
@@ -4707,33 +4770,171 @@ ParserTargetInfo* Parser::GetTargetInfo()
     return parserTargetInfo;
 }
 
-Declaration* Parser::GetDeclarationFromFriend(clang::NamedDecl* FriendDecl)
+ParserResult* Parser::Compile(const std::string& File)
 {
-    Declaration* Decl = WalkDeclarationDef(FriendDecl);
-    if (!Decl) return nullptr;
+    llvm::InitializeAllAsmPrinters();
+    llvm::StringRef Stem = llvm::sys::path::stem(File);
+    Setup(Stem == "Std-symbols");
 
-    int MinLineNumberStart = std::numeric_limits<int>::max();
-    int MinLineNumberEnd = std::numeric_limits<int>::max();
-    auto& SM = c->getSourceManager();
-    for (auto it = FriendDecl->redecls_begin(); it != FriendDecl->redecls_end(); it++)
+    c->getDiagnostics().setClient(new ::DiagnosticConsumer());
+
+    c->getFrontendOpts().Inputs.clear();
+    c->getFrontendOpts().Inputs.push_back(clang::FrontendInputFile(File, clang::Language::CXX));
+
+    const llvm::Triple Triple = c->getTarget().getTriple();
+    llvm::StringRef Dir(llvm::sys::path::parent_path(File));
+    llvm::SmallString<1024> Object(Dir);
+    llvm::sys::path::append(Object,
+        (Triple.isOSWindows() ? "" : "lib") + Stem + ".o");
+    c->getFrontendOpts().OutputFile = std::string(Object);
+
+    llvm::LLVMContext context;
+    auto action = std::make_unique<clang::EmitObjAction>(&context);
+    if (!c->ExecuteAction(*action))
     {
-        if (it->getLocation() != FriendDecl->getLocation())
-        {
-            auto DecomposedLocStart = SM.getDecomposedLoc(it->getLocation());
-            int NewLineNumberStart = SM.getLineNumber(DecomposedLocStart.first, DecomposedLocStart.second);
-            auto DecomposedLocEnd = SM.getDecomposedLoc(it->getEndLoc());
-            int NewLineNumberEnd = SM.getLineNumber(DecomposedLocEnd.first, DecomposedLocEnd.second);
-            if (NewLineNumberStart < MinLineNumberStart)
-            {
-                MinLineNumberStart = NewLineNumberStart;
-                MinLineNumberEnd = NewLineNumberEnd;
-            }
-        }
+        auto res = new ParserResult();
+        HandleDiagnostics(res);
+        return res;
     }
-    if (MinLineNumberStart < std::numeric_limits<int>::max())
+    return 0;
+}
+
+void Parser::Link(const std::string& File, const LinkerOptions* LinkerOptions)
```

Comment:
All this code needs at the very least to be moved to a `Link.cpp` file.

Reply:
This might be useful now what we use a parallel build.

Reply:
Done.